### ISSUE TITLE
net_id rules for arm64 platform driver for hisilicon/hns/hns_enet

### ISSUE THREAD

**Issue by xnox** [systemd/systemd#5894]
### Submission type

  - [ ] Bug report
  - [x] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

Git master & 4.10 kernel

### Used distribution

Ubuntu

### In case of bug report: Expected behaviour you didn't see

Stable interface names for hisilicon platform onboard NICs

### In case of bug report: Unexpected behaviour you saw

Random interface names ranging from eth0..3 to eth4..7, as there are PCI NICs too which get stable names.

### In case of bug report: Steps to reproduce the problem

Boot ARM64 hisilicon board and observe non-stable interface names for the platform hisilicon hns driver.

Currently enabled MAC address based stable names, which are stable but they are not consistent across different servers. Ideally "port 0" would have the same name on all hisilicon servers.

Looking at `/sys/class/net/` listing
```
lrwxrwxrwx  1 root root 0 May  5 10:06 enxa4c64fbcbd47 -> ../../devices/platform/HISI00C2:00/net/enxa4c64fbcbd47
lrwxrwxrwx  1 root root 0 May  5 10:06 enxa4c64fbcbd48 -> ../../devices/platform/HISI00C2:01/net/enxa4c64fbcbd48
lrwxrwxrwx  1 root root 0 May  5 10:06 enxa4c64fbcbd49 -> ../../devices/platform/HISI00C2:02/net/enxa4c64fbcbd49
lrwxrwxrwx  1 root root 0 May  5 10:06 enxa4c64fbcbd4a -> ../../devices/platform/HISI00C2:03/net/enxa4c64fbcbd4a
```
We can observe that NICs are probed in an oder, which matches the MAC address order.

Checking out udevadm database we can see following
```
P: /devices/platform/HISI00C2:02/net/enxa4c64fbcbd49
E: DEVPATH=/devices/platform/HISI00C2:02/net/enxa4c64fbcbd49
E: ID_MM_CANDIDATE=1
E: ID_NET_DRIVER=hns
E: ID_NET_LINK_FILE=/lib/systemd/network/99-default.link
E: ID_NET_NAME_MAC=enxa4c64fbcbd49
E: ID_PATH=platform-HISI00C2:02
E: ID_PATH_TAG=platform-HISI00C2_02
E: IFINDEX=9
E: INTERFACE=enxa4c64fbcbd49
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/enxa4c64fbcbd49
E: TAGS=:systemd:
E: USEC_INITIALIZED=8428125
```

Note the ID_PATH and ID_PATH_TAG.

Reading the hns driver code, we can observe that there are currently two ACPI table matches for this ethernet driver: `HISI00C1` and `HISI00C2`.

Taking ACPI tables dump, and disassembling it the following can be observed:
 * the order of ETHs defined is ETH4, ETH5, ETH0, ETH1
 * port-idx-in-ae matches those suffixes: 0x04, 0x05, Zero, One
 * Device ROCE in the eth-handle defines a list of 6 ETH devices [ETH0, ETH1, Zero, Zero, ETH4, ETH5]
 * the order of the ETHs matches ndev used by kernel
 * there is fallback code in the kernel to read port-id if port-idx-in-ae is not defined

It seems to me that machines could be configured with either HISI00C1 or HISI00C2, and it can have up to 6 NICs configured, of which there are only 4 on this machine.

I'd like to propose to use stable NIC names for these, based on the ACPI Match for HISI00C1 HISI00C2 and the port-idx-in-ae. port-idx-in-ae appears to be stable enumeration of ports, which correctly skips missing ports.

I am not sure if I need to change kernel to expose port-idx-in-ae, and/or if i can just create a new stable nic prefix in udev net_id rules. This feels like an onboard nic, but this kernel driver does not currently populate `phys_port_id`, `phys_port_name`, `phys_switch_id` attributes.

Would be acceptable to propose a patch for hisilicon interface names of the following pattern:

```enh[1..2]p[0..5]```

where h stands for hisilicon / hns driver, 1 or 2 stands for HISI00C1 or HISI00C2, and the number after p stads for port-idx-in-ae.

Or is there some other way, that you would prefer to handle ACPI / device-tree platform ARM NICs?

**Comment by xnox** [comment]
And here are the firmware paths that correspond to the `port-idx-in-ae` attribute
```
$ grep ETH /sys/class/net/enxa4c64fbcbd4*/device/firmware_node/path
/sys/class/net/enxa4c64fbcbd47/device/firmware_node/path:\_SB_.ETH4
/sys/class/net/enxa4c64fbcbd48/device/firmware_node/path:\_SB_.ETH5
/sys/class/net/enxa4c64fbcbd49/device/firmware_node/path:\_SB_.ETH0
/sys/class/net/enxa4c64fbcbd4a/device/firmware_node/path:\_SB_.ETH1
```

**Comment by poettering** [comment]
Since we don't have hw of this type (and have no direct interest in them) we expect patches to add something like this.

Note though that default interface names may be populated by hwdb entries too. If you have a specific device for which stable, friendly names make sense to just add them to the hwdb. For example: if you can recognize this specific kind of device, you could name the devices "front", "back", "red", "green", "one", "two", "three" or any other name you like... 

We haven't really made use of that so far and if we start making use of it we should define some ground rules (i.e. a single hwdb file upstream for these names, and some restrictions on the name, for example by using a common prefix or so)...

**Comment by xnox** [comment]
ACK. I did get hands onto the original ACPI tables code, and got more details about ae-handle and port-idx-in-ae. Turns out that only combination of ae-handle&port-idx-in-ae are unique, yet have no connection to the physical board layout.

The sequence of these devices in ACPI corresponds to the physical board layout. Therefore the best way to name this is to have a driver prefix, and use the driver devnum for this one Thus something like:

<en><h><0..3> where h is short for hisilicon driver.

Would it be OK to propose patch for stable names just for hisilicon driver? Which I think will be similar in simplicity to the recently pulled vio bus? I'm still coordinating with the vendor to get sensible names.
